### PR TITLE
fix #5742 feat(nimbus): implement UI for manual enrollment end 

### LINKED ISSUE #5742: Add end enrollment UI flow
After we switch the backend to end enrollments based on publish_status, we need to add a flow to the UI to trigger it.  It'll following the exact same flow as ending the experiment except setting a different set of states to trigger the tasks.  And we'll need a second button for 'End Enrollment' next to 'End Experiment'.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1349)

### PR DESCRIPTION
**Pull request by lmorchard**
Because:

* we want to add UI to enable manual enrollment end

This commit:

* add isEnrollmentPaused to GQL experiment input for consistency with
  serialized experiment type, mapped to is_paused in Django model

* add support for isEnrollmentPausePending to indicate when the review
  is in progress

* rework useChangeOperationMutation to indicate loading while awaiting
  experiment refetch

* update "rejecting this review" copy to "rejecting the request to
  {actionDescription}"

* refactors handling of strings describing lifecycle review flows, moves
  all the strings into shared constants

* refactors the Settings page and ChangeApprovalOperations components

* adds more exhaustive stories for launch, end enrollment, and end cases
  on PageSummary

* reworks rejection reason display to more properly use old_status and
  old_status_next to determine what kind of review was rejected

* adds more status-based logic to make the correct GQL queries on
  approval / rejection of review

Co-authored-by: Lauren Zugai <lauren@zugai.com>

### CODE REVIEW

**Comment by lmorchard** [comment]
Opening as a draft with work in progress because:

* There's still work to be done around tests & coverage

* I'll be out for another week after the wellness week and someone else might want to grab the baton from me

There's also possibly some some thinking to do about how manual enrollment end is handled on the backend, which could require some further tweaks.

For automated pause, we *do not* set is_paused on the Experimenter side first.

Instead, we set "isEnrollmentPaused" in the Remote Settings record. Then, we wait until it's approved in Remote Settings before setting is_paused=True in Experimenter. That means, on the Experimenter side, we never show that enrollment is paused until it's fully resolved on the Remote Settings side.

For manual pause, we might want to set is_paused=True on the Experimenter side first.

Otherwise, if we follow that same pattern as automated pause, then the only thing that represents a request to pause is status=LIVE / status_next=LIVE. (Because is_paused=False until the very end.)

Could also be a thing to punt to follow-ups, if/when we want other status=LIVE / status_next=LIVE updates other than enrollment ending.

**Comment by lmorchard** [comment]
Rebased against master to pick up the changes to PageSummary and approve / reject handlers

**Comment by lmorchard** [comment]
Looks like I got the storybook to build, so here are some interesting places to look:

* Enrollment active (with new button) - https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/07b3a25088c07c202b64c2ef25f48efe2587c6fa/nimbus-ui/index.html?path=/story/components-summary--enrollment-active
* Enrollment ended (no button) - https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/07b3a25088c07c202b64c2ef25f48efe2587c6fa/nimbus-ui/index.html?path=/story/components-summary--enrollment-ended
* Lots of new launch, pause, and end stories -  https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/07b3a25088c07c202b64c2ef25f48efe2587c6fa/nimbus-ui/index.html?path=/story/pages-summary--live-update-pending-can-review

**Comment by LZoog** [comment]
Responding to [Les' comment](https://github.com/mozilla/experimenter/pull/5852#issuecomment-868895053) + leaving some notes for myself and the reviewer:

Per [Jared's comment here](https://github.com/mozilla/experimenter/pull/5838#issuecomment-874905288), automatically triggered pausing will be removed in favor of requiring manual end enrollment.

> Otherwise, if we follow that same pattern as automated pause, then the only thing that represents a request to pause is status=LIVE / status_next=LIVE. (Because is_paused=False until the very end.)

I think for now, this is OK since there's not any other statuses/flows that use this set and we won't need to differentiate between automatic or manual enrollment requests. That status set (which we're calling `pauseRequested`) does seem a little more fragile than say, `endRequested`, which checks for status=LIVE / status_next=COMPLETE and I could see that perhaps `is_paused` (`isEnrollmentPaused`) could potentially do a little more in the FE for us but we _are_ using `isEnrollmentPaused` in conjunction with `pauseRequested` to conditionally show the `EndEnrollment` component. 🤔 FWIW we also aren't using it for end enrollment RS timeouts either since we just check for `timeout` on the experiment.

If we do have another flow later that needs status=LIVE / status_next=LIVE, we should be able to alter `pauseRequested` on the FE fairly easily. With all of this said, I think this PR mostly just needed some hard squinting at, test fixes and additions, warning fixes, and a couple follow up issues filed, all of which I've got underway. Just wanted to dump my thoughts here before finishing those up.

File: app/experimenter/experiments/models/nimbus.py
Comment (LZoog) [inline app/experimenter/experiments/models/nimbus.py:157]: This comment was Les' and I'm not sure what else we need here?

File: app/experimenter/nimbus-ui/src/components/PageSummary/index.test.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/PageSummary/index.test.tsx:313]: I added these `findByText` tests to double check the test was actually testing the correct flow.

File: app/experimenter/nimbus-ui/src/components/Summary/index.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/Summary/index.tsx:53]: I'm guessing this ties in with [this TODO](https://github.com/mozilla/experimenter/pull/5852/files#r666457588) where at the moment I'm not sure what else we need

File: app/experimenter/nimbus-ui/src/components/Summary/index.test.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/Summary/index.test.tsx:130]: FWIW we should end up using shared common props for flows when we do #5851. I didn't pull in [the mocks from PageSummary](https://github.com/mozilla/experimenter/pull/5852/files#diff-a22e33c18e87900a21ac78b38246317600edd36c1cc9885b6d322d4e1fb170f5R114) since we probably need to pull those shared props into a separate file at a higher level which would require some more refactoring and make this PR even longer, plus that's what #5851 is for.

**Comment by LZoog** [comment]
To follow up here, yesterday I pushed test fixes and additions plus console warnings and a couple of other tweaks but left BE test fixes since we need to wait until there's a PR closing #5739.

I also already rebased/squashed the commits because of habit and I didn't realize until after pushing that this PR is probably going to stick around by the time @lmorchard gets back. 😅

File: app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/RejectionReason.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/RejectionReason.tsx:40]: Oh @lmorchard I see what you were saying in planning, #5927 covers the other side of this, "You are rejecting..." instead of "The request to". Could easily be done in this PR or in that issue separately.

File: app/experimenter/experiments/models/nimbus.py
Comment (lmorchard) [reply app/experimenter/experiments/models/nimbus.py:157]: This comment points toward basically [what I mentioned in this PR comment](https://github.com/mozilla/experimenter/pull/5852#issuecomment-868895053) - we don't set `is_paused=True` here, just a LIVE->LIVE status change that gets interpreted as a request to pause.

**Comment by lmorchard** [comment]
Going to pick this back up tomorrow - PR #5935 will result in a totally different backend situation along with requiring we send `is_paused={True,False}` from the frontend.

So, some reworking will be required. But, hopefully not more than ripping out the backend changes in this PR and just updating the GQL queries on the frontend though.

**Comment by lmorchard** [comment]
Alright, I think this is ready for review now. Stripped out the defunct backend changes, added some to allow mutating isEnrollmentPaused / is_paused. Also tossed in a fix for #5927 as a quick bonus. The `Co-authored-by` might disqualify @LZoog from reviewing, since she helped carry this PR for the week I was out and wrapped up test coverage.

File: app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/RejectionReason.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/RejectionReason.tsx:40]: Added a fix for this https://github.com/mozilla/experimenter/pull/5852/files#diff-c86b7877156b25769ecb6879ff40aef78f66ac82fa8ff115b6e679b6870a92b9R54-R57

**Comment by jaredlockhart** [comment]
Clicking the End Experiment button sends an API request but doesn't update the UI.  Refreshing the page kicked it over, but looks like we're missing something there.

**Comment by jaredlockhart** [comment]
Approve dialog missing some text for End Enrollment:

Screen Shot 2021-07-14 at 10 22 58 AM

**Comment by jaredlockhart** [comment]
Clicking reject also seems to not update UI:

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
Hit a bunch of snags up front, added some gifs.  I'll stop testing there for now and look again when I can get further through the flow.

**Comment by LZoog** [comment]
> Clicking the End Experiment button sends an API request but doesn't update the UI. Refreshing the page kicked it over, but looks like we're missing something there.

@lmorchard FWIW I think that's taken care of in #5939 if you rebase

**Comment by lmorchard** [comment]
I see what happened with the missing text & GQL query: I added in an `isEnrollmentPaused` condition to `status.pauseRequested` and that broke things. Turns out `isEnrollmentPaused` is now derived from `is_paused_published` rather than `is_paused`, and that only turns true after review completion. Might need to add another property to represent `is_paused` vs `is_paused_published`

**Comment by lmorchard** [comment]
> @lmorchard FWIW I think that's taken care of in #5939 if you rebase

Hmm, that only takes care of the end experiment UI, not this new UI. Need to copy that over / see if it can be more generalized

File: app/experimenter/experiments/tests/api/v5/test_queries.py
Comment (lmorchard) [inline app/experimenter/experiments/tests/api/v5/test_queries.py:760]: Figured I might as well try exhausting all the cases here to make sure the pending vs published pause states worked as expected.

File: app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.tsx:68]: This seems to better ensure that the UI reflects that something's still in-flight for the refetch after the mutation

File: app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx:37]: Reworking these tests to use [`await waitForNextUpdate()`](https://react-hooks-testing-library.com/usage/advanced-hooks#async) seems to exercise the functionality better. But I suspect there might be an intermittent race condition in here still that I haven't been able to flush out

File: app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx:61]: Seemed like `customMocks` was never used in these tests?

File: app/experimenter/nimbus-ui/src/components/AppLayoutWithExperiment/index.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/components/AppLayoutWithExperiment/index.tsx:23]: Not sure where else we're using `refetch` - this change doesn't seem to have broken any other tests, but we might need to update this type elsewhere eventually. Technically the refetch from Apollo *is* an async function and not `() => void`, but we didn't have that plumbed through for use in `useChangeOperationMutation`

File: app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx:59]: This feels a little dirty, but seems like a decent way to predictably walk through the states around refetch.

File: app/experimenter/experiments/api/v5/types.py
Comment (lmorchard) [inline app/experimenter/experiments/api/v5/types.py:228]: This combined with status=LIVE / status_next = LIVE should give us all the hints needed on the frontend

**Review by lmorchard — COMMENTED** [review]
This PR has become monstrous, but I think it's got some needed refactorings. Commenting on some additions that should fix the UI with respect text, loading indication, and using the correct GQL mutation for end enrollment

**Comment by jaredlockhart** [comment]
@lmorchard If you feel like there's refactors in here that are necessary but don't actually implement the new UI directly you could pull them out into separate more isolated PRs, review and land those first, and then rebase this PR to just contain the actual logic to implement the new UI.  Up to you.

**Comment by lmorchard** [comment]
> If you feel like there's refactors in here that are necessary but don't actually implement the new UI directly you could pull them out into separate more isolated PRs, review and land those first, and then rebase this PR to just contain the actual logic to implement the new UI. Up to you.

No, it's all pretty much inseparable at this point. The refactors were necessary to get the shared logic working between the now three review cases we have

**Comment by lmorchard** [comment]
Filed #9852 to capture a refactor idea for `useChangeOperationMutation` and keep this PR from getting bigger.

File: app/experimenter/nimbus-ui/src/lib/constants.ts
Comment (jodyheavener) [inline app/experimenter/nimbus-ui/src/lib/constants.ts:73]: Nit: I might phrase these as "X Requested", but def feel free to ignore :P

**Review by jodyheavener — COMMENTED** [review]
Stared at this for a bit and had a few thoughts, but nothing major. I was able to go through the flow from start to end (as well as rejections) and didn't get hung up on anything. 🥳

File: app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx:56]: I can't find an issue (which I thought existed) but I think we've been pushing to be more verbose and explicit about these messages. Might be worth a follow-up issue to revisit though

File: app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx:23]: Could be worth a follow-up issue, e.g. to combine these two buttons into a component side-by-side with state to display the confirmation form mutually exclusively for one or the other. Seems a decent enough chunk of work to avoid making this PR bigger

File: app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/index.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/index.tsx:31]: Yeah, I keep thinking some kind of Context with experiment could be a good refactor, maybe also take the place of that functional child pattern we use for all the experiment-using pages

File: app/experimenter/experiments/api/v5/types.py
Comment (jaredlockhart) [reply app/experimenter/experiments/api/v5/types.py:228]: Wait explain this.  Shouldn't you just be able to infer all this based on publish_status? Why do you need to check the difference between the pause and published pause state?

File: app/experimenter/nimbus-ui/src/lib/experiment.ts
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/lib/experiment.ts:39]: The only LIVE>LIVE update we have right now is pausing so that should be sufficient to know we're in a pausing flow?

Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/lib/experiment.ts:39]: Wanted to be specific here and only catch the pending end enrollment case

File: app/experimenter/experiments/api/v5/types.py
Comment (lmorchard) [reply app/experimenter/experiments/api/v5/types.py:228]: Can't infer specifically that this is a pending change to end enrollment, and I wanted to be specific https://github.com/mozilla/experimenter/pull/5852#discussion_r670715048

File: app/experimenter/nimbus-ui/src/lib/experiment.ts
Comment (jaredlockhart) [reply app/experimenter/nimbus-ui/src/lib/experiment.ts:39]: Oh I see what you mean.  I think we'll end up having to do a lot more infrastructure before we do any non pause live updates, so we shouldn't really need to worry about it for now.

File: app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx
Comment (jaredlockhart) [reply app/experimenter/nimbus-ui/src/components/Summary/EndEnrollment.tsx:24]: Yeah good point, interesting thought.  You're right that they're pretty heavily guarded by the multiple review steps.  I'd say leave it and if it starts tripping people up we can revisit it, but good thing to point out. 👍

File: app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx
Comment (jaredlockhart) [reply app/experimenter/nimbus-ui/src/hooks/useChangeOperationMutation.test.tsx:37]: Circle reveals all.

**Review by jaredlockhart — APPROVED** [review]
Tested locally, everything went off without a hitch.  Looks great, works great, thanks @lmorchard ! 🎉 🎉 🎉 🎉

**Comment by lmorchard** [comment]
Alright, I'm gonna hit the merge button to get this monster landed, then file a few follow up issues that are hopefully quick maintenance sized things

**Comment by lmorchard** [comment]
Tried to capture some of @jodyheavener's feedback in follow ups:

* Refining the end action buttons - https://github.com/mozilla/experimenter/issues/5970
* Thinking about Context to pass experiment around - https://github.com/mozilla/experimenter/issues/5971

**Pull request merged**
